packages/@uppy/companion-client/src/RequestClient.js

```diff
@@ -222,179 +240,295 @@ export default class RequestClient {
-          socket.send('pause', {})
-          socket.send('resume', {})
+        const onFileRemove = (targetFile) => {
+          if (!capabilities.individualCancellation) return
+          if (targetFile.id !== file.id) return
+          socketSend('cancel')
+          socketAbortController?.abort?.()
+          this.uppy.log(`upload ${file.id} was removed`, 'info')
+          resolve()
         }
-      })
 
-      socket.on('progress', (progressData) => emitSocketProgress(this, progressData, file))
-
-      socket.on('error', (errData) => {
-        const { message } = errData.error
-        const error = Object.assign(new Error(message), {
-          cause: errData.error,
-        })
-
-        // If the remote retry optimisation should not be used,
-        // close the socket—this will tell companion to clear state and delete the file.
-        if (!this.opts.useFastRemoteRetry) {
-          // Remove the serverToken so that a new one will be created for the retry.
-          this.uppy.setFileState(file.id, {
-            serverToken: null,
-          })
-        } else {
-          socket.close()
+        const onCancelAll = ({ reason }) => {
+          if (reason === 'user') {
+            socketSend('cancel')
+          }
+          socketAbortController?.abort?.()
+          this.uppy.log(`upload ${file.id} was canceled`, 'info')
+          resolve()
+        };
+
+        const onFilePausedChange = (targetFileId, newPausedState) => {
+          if (targetFileId !== file.id) return
+          pause(newPausedState)
         }
 
-        this.uppy.emit('upload-error', file, error)
-        queuedRequest.done()
-        reject(error)
-      })
-
-      socket.on('success', (data) => {
-        const uploadResp = {
-          uploadURL: data.url,
+        const onPauseAll = () => pause(true)
+        const onResumeAll = () => pause(false)
+
+        this.uppy.on('file-removed', onFileRemove)
+        this.uppy.on('cancel-all', onCancelAll)
+        this.uppy.on('upload-pause', onFilePausedChange)
+        this.uppy.on('pause-all', onPauseAll)
+        this.uppy.on('resume-all', onResumeAll)
+
+        removeEventHandlers = () => {
+          this.uppy.off('file-removed', onFileRemove)
+          this.uppy.off('cancel-all', onCancelAll)
+          this.uppy.off('upload-pause', onFilePausedChange)
+          this.uppy.off('pause-all', onPauseAll)
+          this.uppy.off('resume-all', onResumeAll)
         }
 
-        this.uppy.emit('upload-success', file, uploadResp)
-        queuedRequest.done()
-        socket.close()
-        resolve()
-      })
-
-      queuedRequest = queue.run(() => {
-        if (file.isPaused) {
-          socket.send('pause', {})
-        } else {
-          socket.open()
-        }
+        signal.addEventListener('abort', () => {
+          socketAbortController?.abort();
+        })
 
-        return () => {}
+        createWebsocket()
       })
-    })
+    } finally {
+      removeEventHandlers?.()
```

Comment:
is finally always reached? Also in the case of `cancel-all`?

Reply:
cancel-all calls onCancelAll which calls resolve https://github.com/transloadit/uppy/blob/6d813c91e5a40882af479ed612049d775596174d/packages/%40uppy/companion-client/src/RequestClient.js#L499C16-L499C16 which should hit the finally block yes because we `try { return await new Promise((resolve, reject)`